### ISSUE TITLE
training error

### ISSUE THREAD

**Issue by xgxgnpu** [hanfengzhai/PIDOC#1]
您好，我运行您的benchmark_main代码时，发现代码中的优化器部分少了for循环，还有优化过程loss也没有下降，请问是什么原因呢？目前您提供的代码好像无法复现您的论文结果，是不是您上传的代码版本没有更新呢？

**Comment by hanfengzhai** [comment]
您好，实在抱歉，我做了double-check，做以下补充：
1. 要重现论文（图4的benchmark analysis），训练数据要为"train.mat"。对我的粗心感到抱歉！
2. benchmark 的 train_10.mat是加入了噪音的训练数据，是我们回复reviewer对于噪音的质疑引入的数据。您如果plot(t,x_10)可以看到效果，该数据在我们的文章中没有出现和提及，纯属是我们的补充数据，对于代码上传粗心深感抱歉。对于train_10，您可以点击https://hanfengzhai.net/data/github_response.png 查看。
3. 损失函数在训练一定次数之后不下降是正常现象。如果您仔细看文章中的 图5 & 图6 & 图7，我们实验中损失函数在一定次数也不下降了。原因为：PIDOC的本质目的不是拟合训练数据（不是用NN做数据拟合预测），而是enforce控制信号（这是PIDOC的核心idea），所以和训练集的差会使得损失数量级较大。但是我们还是可以expect在初期会有损失（目标）函数的下降的（i.e.从一个很大的数值逐渐降低）。在这里面损失函数在这里相当于多目标优化把多个目标函数simply乘上一个拉格朗日算子加到了一起，是多目标优化中最常见也是最简单的做法。因为两个目标函数（MSE_NN和MSE_D）本质上就是不同的，所以在这里不可能出现像传统神经网络那样损失函数降到一个特别小的数值的情况。详情请看Section 3.1第5和6自然段。
4. 对于您指出的优化器中的for循环（如果我对您的问题理解正确的话），我们没有用PINN原文中Iteration去拟合任何参数，在这里我们simply想做一个控制信号的enforcement，所以我们只是用了最简单的目标函数的优化来测试信号encoding，因为二者的目的完全不一样。如果我对您的问题理解错误，请指出谢谢！
5. 如果您要复现图5，需要手动调38、39和48、49行的Lambda（在benchmark_main.py中的2* 和 /2）；如果您想要复现图7，需要改变代码107行的初始位置；如果要复现图7 & 8，需要完全改变系统数据输入，在我们的data中有对应的非线性训练集（我们也在文中明确指出PIDOC在对强非线性系统结果并不理想，从图7 & 8也可以看出来，in fact我们也做了和传统控制理论的结果对比，nonlinear feed-forward open-loop control的结果在mu不大时是要好于PIDOC的，不过close loop对比下PIDOC是要好于传统控制算法的，不过我们并没有把结果放上去）；如果您要复现图9，需要手动调网络结构（改变第145行），如果想要复现图10，需要调拉格朗日约束大小（第48行）。我理解以上过程很繁琐而且我的代码开源非常不全面，对此深感抱歉！
同时，我们也做了（目前正在refine）PIDOC和传统控制方法（feed-forward & feedback & combined）的对比实验，但并没有在论文中放结果。如果您对控制对比的结果感兴趣，我们也可以给您提供Simulink文件（请提供邮箱）。如果您有进一步问题，请指出，谢谢！
感谢您对该工作感兴趣，我对代码上传不足深感抱歉。我将在近一个月内把对应的每张图都写出对应的代码并上传！
再次对数据上传不足和粗心感到抱歉！如果有进一步问题请指出！

**Comment by hanfengzhai** [comment]
> 您好，我运行您的benchmark_main代码时，发现代码中的优化器部分少了for循环，还有优化过程loss也没有下降，请问是什么原因呢？目前您提供的代码好像无法复现您的论文结果，是不是您上传的代码版本没有更新呢？

如果您只是想要尽快得到图4的benchmark analysis，请直接refer到：https://github.com/hanfengzhai/PIDOC/blob/main/vanderPol/vanderPol_Fig4_benchmark.ipynb ，得到控制后的数据prediction.txt您可以使用后处理得到数值求导然后画phase portrait就可以得到图4。后续我会尽快补充其余部分代码，谢谢。

**Comment by hanfengzhai** [comment]
> 您好，我运行您的benchmark_main代码时，发现代码中的优化器部分少了for循环，还有优化过程loss也没有下降，请问是什么原因呢？目前您提供的代码好像无法复现您的论文结果，是不是您上传的代码版本没有更新呢？

运行代码后您应该得到类似这样的controlled scheme：https://hanfengzhai.net/data/response_github_2.png 。对应图4的benchmark。

**Comment by xgxgnpu** [comment]
非常感谢您耐心详细的解答！您的想法设计非常巧妙，结果出乎意料！

您的5条回复已经解答了我很多疑惑了，我又重新拜读了您的文章和跑了一些实验，也希望您能提供相应的Simulink文件以便进行对比实验，我的邮箱xiongxiongnwpu@mail.nwpu.edu.cn

然后，我还有些疑惑和个人直观理解，也不知道合不合理。

1、对于MSE_D残差项的设计，您的文章中写了0阶导数加上2阶导数，我又测试了其他组合（0+1+2，1+2，0，1，2等等），似乎只有您的这个搭配能让NN产生比较好的周期解。请问这个有什么原因，原理或者依据吗？
self.control = self.x_res + self.x_dd_res

2、我对于您paper中的想要实现的基本目标的直观理解。 在van der Pol 方程中，关键系统阻尼参数mu，当mu取不同值时，系统呈现不同的非线性行为。
mu=0， 系统对应周期解（sin, cos）；mu>0,系统进入 a limit cycle。您的目标是要把极限环或者更复杂的混沌行为控制到周期解上（如果没理解错的话），那就相当于把mu > 0对应的方程极限环解的数据控制到mu =0的方程周期解的数据。

您的paper中发现，对于非线性阻尼系数mu（0，1，3，5，7，9），当mu越大，系统的非线性越强，PIDOC控制的效果越不理想。我感觉这个跟您的数据相似度有关。当mu=0,解是周期的，随着mu增大，解越来越偏离周期解，这里面似乎有个距离的度量（比如，余弦相似度，mu = 0 对应的解其实就是想要得到的目标信号，mu>0 对应的解就是您的原始信号，mu =1,3,5,7,9对应的方程数据和mu  = 0对应的方程数据的距离或者相似度不同）。    

 也就是说，目标信号和原始信号越接近（mu越靠近0），对于multi-task learning越容易实现，PINN本质上就是一个多任务学习优化问题，对应您的第4条回复（因为两个目标函数（MSE_NN和MSE_D）本质上就是不同的，所以在这里不可能出现像传统神经网络那样损失函数降到一个特别小的数值的情况）。我个人感觉正是因为MSE_NN和MSE_D）本质上就是接近的或者说相似的，所以才能实现控制。
 self.loss = 1 * tf.reduce_sum(tf.square(self.control)) + tf.reduce_sum(tf.square(self.ICs)) +  tf.reduce_sum(tf.square( self.x_tf - (self.x_pred) )) 


mu = 0就对应MSE_NN，而mu>1就对应MSE_D。mu越接近0，MSE_NN和MSE_D越相似，越容易控制；mu越远离0，MSE_NN和MSE_D差异越大，越难控制（主要问题应该还是多个不同Loss的landscape不同，多目标优化难题（参考论文Characterizing possible failure modes in physics-informed neural networks，https://arxiv.org/abs/2109.01050）

**Comment by xgxgnpu** [comment]
另外，对于figure5，with five different desired trajectories Λ = 1, 2, . . . , 5，对于相同的原始信号，施加不同振幅Λ的控制目标信号似乎都能实现较好的控制效果，这里不同振幅的控制信号和原始信号的拓扑结构应该是相似的，某种意义下的距离度量是相似的，是否可以在loss中再扩充一项原始信号和控制信号的距离或者相似性度量

**Comment by xgxg1314** [comment]
我做了进一步的实验，发现当mu=9时，在训练的过程中，也能学到不一样的解，修改后的测试代码见https://github.com/xgxg1314/PINN_control/blob/main/benchmark_main.ipynb。当L-BFGS-B训练轮数=200，可以得到跟论文中不一样的结果（此时也许也可以得到更好的学习或者控制效果）https://github.com/xgxg1314/PINN_control/blob/main/trainingResult_mu9_epoch200.png（当然此时还没有收敛）；当L-BFGS-B训练轮数=300（>300）时，可以得到论文中结果https://github.com/xgxg1314/PINN_control/blob/main/trainingResult_mu9_epoch300.png。

按照神经网络的频率学习法则，NN会优先收敛到低频信息，学会低频信息后再学习高频信息。这里的低频信息是MSE_D中包含的周期性，也就是训练200轮后，NN优先学到了任务MSE_D，继续训咯到300轮，NN会逐步收敛到MSE_NN中包含的高频信息（mu=9，任务MSE_NN中包含很多高频成分）。也就是说，任务MSE_D和任务MSE_NN这两个学习任务，其中谁是低频的，就会优先收敛谁的，对于论文中的控制目标，比较关键的可能是频率，也就是self.x_control = 2 * tf.math.sin(self.t_tf)这里的频率，这是您的paper中没有讨论和忽略的一点。如果选择合适的控制频率，应该就可以学到控制更加复杂（mu比较大）系统非线性。

我目前感兴趣的是multi-task learning 和optimization，看到您的实验结果非常有意思，所以就想深入了解一下。当NN同时学习2个或者多个学习任务时，如何促进NN同时学好多个任务或者只促进学好某一个任务。

您的这个问题应该算同时学习两个任务，MSE_NN和MSE_D，而目标则是倾向于学好任务MSE_D（MSE_NN作为辅助任务促进主任务MSE_D的学习），不知道我对您的目标理解是否合适，希望有时间能深入讨论一下，谢谢！

**Comment by xgxgnpu** [comment]
我做了进一步的实验，发现当mu=9时，在训练的过程中，也能学到不一样的解，修改后的测试代码见

https://github.com/xgxg1314/PINN_control/blob/main/benchmark_main.ipynb

当L-BFGS-B训练轮数=200，可以得到跟论文中不一样的结果（此时也许也可以得到更好的学习或者控制效果）https://github.com/xgxg1314/PINN_control/blob/main/trainingResult_mu9_epoch200.png

（当然此时还没有收敛）；当L-BFGS-B训练轮数=300（>300）时，可以得到论文中结果https://github.com/xgxg1314/PINN_control/blob/main/trainingResult_mu9_epoch300.png

按照神经网络的频率学习法则，NN会优先收敛到低频信息，学会低频信息后再学习高频信息。这里的低频信息是MSE_D中包含的周期性，也就是训练200轮后，NN优先学到了任务MSE_D，继续训咯到300轮，NN会逐步收敛到MSE_NN中包含的高频信息（mu=9，任务MSE_NN中包含很多高频成分）。也就是说，任务MSE_D和任务MSE_NN这两个学习任务，其中谁是低频的，就会优先收敛谁的，对于论文中的控制目标，比较关键的可能是频率，也就是self.x_control = 2 * tf.math.sin(self.t_tf)这里的频率，这是您的paper中没有讨论和忽略的一点。如果选择合适的控制频率，应该就可以学到控制更加复杂（mu比较大）系统非线性。

我目前感兴趣的是multi-task learning 和optimization，看到您的实验结果非常有意思，所以就想深入了解一下。当NN同时学习2个或者多个学习任务时，如何促进NN同时学好多个任务或者只促进学好某一个任务。

您的这个问题应该算同时学习两个任务，MSE_NN和MSE_D，而目标则是倾向于学好任务MSE_D（MSE_NN作为辅助任务促进主任务MSE_D的学习），不知道我对您的目标理解是否合适，希望有时间能深入讨论一下，谢谢！

**Comment by xgxg1314** [comment]
神经网络的Frequency Principle频率法则见上交大许志钦[https://zhuanlan.zhihu.com/p/340646831]

**Comment by xgxgnpu** [comment]
神经网络的Frequency Principle频率法则见上交大许志钦https://zhuanlan.zhihu.com/p/340646831

**Comment by hanfengzhai** [comment]
您好，非常感谢！您的评论太棒了！至于MSE_D，这也是我们在实验中遇到的一个问题，我们也无法提供非常严谨的数学证明来说明控制信号的作用。实在抱歉最近课业较重，无法及时回复。对于您的评论，我附加了材料给您发了邮箱。

**Comment by hanfengzhai** [comment]
> 神经网络的Frequency Principle频率法则见上交大许志钦https://zhuanlan.zhihu.com/p/340646831

您好，我之前已经跟进附件材料给您发了邮件，不知您是否收到。若没收到或进了垃圾邮件我可以再给您发一遍，谢谢！

**Comment by xgxgnpu** [comment]
我刚看了邮箱，不好意思，最近几天回家了一趟，网络不方便，谢谢您的分享，我会仔细看一下您文章的审稿人意见和回复意见，希望可以follow up

**Comment by xgxgnpu** [comment]
不知道您考虑过控制信号频率的影响没？信号的频率对于NN的收敛性影响很大
